README.md

```diff
@@ -2,4 +2,61 @@
 
 A multi-threaded implementation of a PromQL Query Engine based on the [Volcano/Iterator model](https://paperhub.s3.amazonaws.com/dace52a42c07f7f8348b08dc2b186061.pdf).
 
-The project is currently under development.
+The project is currently under active development.
+
+## Design
+
+At the beggining of a PromQL query execution, the query engine computes a physical plan consisting of multiple independent operators, each responsible for calculating one part of the query expression.
+
+Operators are assembled in a tree-like structure with every operator calling `Next()` on its dependants until there is no more data to be returned. The result of the `Next()` function is a _column vector_  (also called a _step vector_) with elements in the vector representing samples with the same timestamp from different time series.
```

Comment:
I wonder if we could mention IF this step vector contains ALL step series OR a partial view of series. As I see the design permits both, which might offer better scalability (no need for pulling all data for query in memory).

Reply:
Right now we populate the step vector with all series. We could play with the partial mode as well and if that is possible to do, then memory management will also become better since we will be able to use fixed size vectors that are sent forward when full.